File: day11/main.cpp
// https://adventofcode.com/2024/day/11

#include <array>
#include <cmath>
#include <iostream>
#include <map>
#include <vector>

#include "../utility/helpers.hpp"

using IntType = unsigned long long;

[[nodiscard]] int getNumberOfDigits(IntType n) {
  return n == 0 ? 1 : std::floor(std::log10(n)) + 1;
}

IntType stringToInt(const std::string &str) {
  std::istringstream iss(str);
  IntType val;
  iss >> val;
  return val;
}

[[nodiscard]] std::array<IntType, 2> EvenDigitsRule(IntType n) {
  std::string asStr = std::to_string(n);
  int divisionPoint = asStr.size() / 2;

  IntType half1 = stringToInt(asStr.substr(0, divisionPoint));
  IntType half2 = stringToInt(asStr.substr(divisionPoint));

  return {half1, half2};
}

[[nodiscard]] IntType ElseRule(IntType n) { return n * 2024; }

IntType countCached(IntType stone, IntType steps,
                    std::map<std::pair<IntType, IntType>, IntType> &cache) {
  auto key = std::make_pair(stone, steps);
  if (cache.find(key) != cache.end())
    return cache[key];

  if (steps == 0)
    return 1;
  if (stone == 0)
    return countCached(1, steps - 1, cache);

  IntType res{};

  if (getNumberOfDigits(stone) % 2 == 0) {
    auto [half1, half2] = EvenDigitsRule(stone);
    res = countCached(half1, steps - 1, cache) +
          countCached(half2, steps - 1, cache);
  } else {
    res = countCached(stone * 2024, steps - 1, cache);
  }

  return cache[key] = res;
}

int main() {
  std::string txt = getRawText("input.txt");
  std::vector<IntType> stones;

  {
    std::istringstream iss(txt);
    IntType n;
    while (iss >> n)
      stones.push_back(n);
  }

  std::map<std::pair<IntType, IntType>, IntType> cache;
  for (IntType blink : {25, 75}) {
    IntType sum{};
    for (IntType stone : stones) {
      sum += countCached(stone, blink, cache);
    }

    std::cout << sum << '\n';
  }
}

File: day11/src/main.rs
use std::{collections::HashMap, io, str::FromStr};

use utility::get_raw_text;

type IntType = u128;

fn num_len(n: IntType) -> usize {
    match n {
        0 => 1,
        _ => n.ilog10() as usize + 1,
    }
}

fn string_to_int(s: &str) -> IntType {
    IntType::from_str(s).unwrap()
}

fn even_digits_rule(n: IntType) -> [IntType; 2] {
    let as_str = n.to_string();
    let division_point = as_str.len() / 2;

    let half1 = string_to_int(&as_str[..division_point]);
    let half2 = string_to_int(&as_str[division_point..]);

    [half1, half2]
}

fn else_rule(n: IntType) -> IntType {
    n * 2024
}

fn count_cached(
    stone: IntType,
    steps: IntType,
    cache: &mut HashMap<(IntType, IntType), IntType>,
) -> IntType {
    let key = (stone, steps);

    if let Some(&cached) = cache.get(&key) {
        return cached;
    }

    if steps == 0 {
        return 1;
    }

    if stone == 0 {
        return count_cached(1, steps - 1, cache);
    }

    let res = if num_len(stone) % 2 == 0 {
        let [half1, half2] = even_digits_rule(stone);
        count_cached(half1, steps - 1, cache) + count_cached(half2, steps - 1, cache)
    } else {
        count_cached(else_rule(stone), steps - 1, cache)
    };

    cache.insert(key, res);
    res
}

fn main() -> io::Result<()> {
    let txt = get_raw_text("input.txt")?;
    let mut stones = Vec::new();

    for word in txt.split_whitespace() {
        stones.push(string_to_int(word));
    }

    let mut cache = HashMap::new();
    for blink in [25, 75] {
        let mut sum = 0;
        for &stone in &stones {
            sum += count_cached(stone, blink, &mut cache);
        }
        println!("{}", sum);
    }

    Ok(())
}
